### ISSUE TITLE
What platforms support MFX_FRCALGM_FRAME_INTERPOLATION?

### ISSUE THREAD

**Issue by OpenSourceAnarchist** [Intel-Media-SDK/MediaSDK#2065]
I know Haswell (gen 4) supports this algorithm, but are modern CPUs still supported? Additionally, there were reports (see this unresolved thread: https://software.intel.com/en-us/forums/media/topic/538086) that the algorithm simply duplicates frames without interpolating them. What is the best way to ensure frame interpolation occurs?

I was hoping Intel's frame interpolation approach would work as an alternative to AMD Fluid Motion and CUDA-only approaches that are not compatible with my current hardware. Thanks for any info!

**Comment by dmitryermilov** [comment]
As far as I know MediaSDK VPP never supported interpolation in FRC, it always was a dummy frame duplication/dropping.
@FurongZhang , can you please clarify from HW side?

**Comment by OpenSourceAnarchist** [comment]
Thanks for the clarification. Since Intel Media Server Studio is closed-source, I suppose it would be difficult to copy that functionality to MediaSDK, if it was ever supported. Apparently this was a hardware-only feature and was never implemented as a sw algorithm.

**Comment by OpenSourceAnarchist** [comment]
I hate to keep commenting, but according to the article here: https://software.intel.com/en-us/articles/frame-rate-conversion-frc-in-intel-media-sdk, the interpolation algorithm was a part of MediaSDK, at least at one point. I'm not sure if this helps, but it was exposed as of version 1.6 of the API, and according to the 2014 Developer's Guide (see https://jp.xlsoft.com/documents/intel/media/2014/Intel_Media_Developers_Guide_0.pdf), Section 4.10.4, "Many implementations of Intel® Iris™ Pro Graphics, Intel® Iris™ Graphics and Intel® HD Graphics (4200+ Series) allow support the creation of interpolated frame content by using the MFX_FRCALGM_FRAME_INTERPOLATION algorithm," so it was definitely enabled at some point.

**Comment by dmitryermilov** [comment]
Thanks @OpenSourceAnarchist . I checked old internal documentation and MSS code. Indeed, it supported FRC with interpolation (at least on HSW). Let's wait for Furong's reply.

**Comment by OpenSourceAnarchist** [comment]
It's been a few months... I was just wondering if there's been any updates from frame interpolation in MediaSDK. The Media SDK Developer Reference still mentions various options related to frame rate interpolation and I haven't tested anything recently. Any news from @FurongZhang?

**Comment by FurongZhang** [comment]
@OpenSourceAnarchist , I read your comments in this thread. Your understanding is correct and you did a great work for digging out the history. Yes, we once supported Frame Rate Conversion on some platforms from graphics driver side. But from our KBL, graphics driver will not support FRC anymore. We don't know any typical usage, hence we stop the investment on this feature. Now, from MSDK side, just support frame duplication and drop for FRC, no interpolation supported. 
May I know why you are so interested in this feature? Thanks.

**Comment by OpenSourceAnarchist** [comment]
@FurongZhang There are a number of existing frame interpolation options, however most remain inaccessible or mostly impractical. 

Almost all modern TVs have a dedicated chip to perform frame interpolation in real-time, but this cannot be used to encode videos and is hardly configurable.

Software like [SVP](https://www.svp-team.com/) is useful for this purpose, as well as real-time frame interpolation, but it is not open-source and has the problem of needing to use the CPU or OpenCL for the frame interpolation algorithm which requires copying-back frames between the CPU and GPU.

FFmpeg's libavfilter library has a video filter called minterpolate, but this is terribly slow as it is single-threaded and works only on the CPU.

Any neural-net options, such as DAIN, sepconv, or NVIDIA's Super-SloMo, are nowhere near the speed or quality of SVP with the right settings. Both AMD and NVIDIA have existing frame interpolation support in their graphics drivers, but they are limited to gaming (NVIDIA) or older video cards and videos with a frame rate of 25fps or 30fps (AMD). If Intel could create an open-source frame interpolation method that works on the GPU, I can see a large number of high-frame rate enthusiasts switching over from using SVP, custom AviSynth scripts, or other undesirable options to MediaSDK for both real-time viewing and encoding of <60fps content to 60+fps content. It may not seem that useful, but there is a very dedicated community to viewing films, anime, and TV shows in a higher frame rate than the original. It also has a very important use-case for slowing down videos and having the result appear smooth.

I am hoping that Intel can take advantage of the lack of quality options for this feature and effectively blow the competition, which consists of mostly non-free software, out of the water. I imagine that some of the motion estimation code for h.264 and h.265 could be re-used to get motion vectors for pixels, and referring to the best frame rate conversion and motion estimation library, mvtools, would be useful to see an exiting C implementation of the algorithms needed for this feature.

This is not a small feature to ask for, but there are many potential users—researchers and enthusiasts—that would benefit from this option in MediaSDK. This was a longer post, but I have experimented a lot with frame interpolation and just wish there was a better GPU-accelerated option available for users!

**Comment by FurongZhang** [comment]
@OpenSourceAnarchist I got it. I am very interested in this topic. I am curious who you are and what you are working on? If it is not good to put that here, could you please write me an email? Thanks a lot in advance!

**Comment by OpenSourceAnarchist** [comment]
I'm an undergrad at the University of Pennsylvania, but really I'm just just a high frame rate enthusiast. A few years ago I discovered how accurate and preferable (to some people, at least) frame interpolation can be when watching many types of media, but unfortunately there are only a few software options, and even though SVP is free for Linux users, it is difficult to get working on many distros other than Arch Linux. I've been looking for an open-source alternative for a long time and the closest I have found is an old Avisynth script that is only uses the CPU and does not produce acceptable real-time results. When I saw frame interpolation support in the MediaSDK manual a few months ago, I was elated to see an open-source, cross-platform solution that is hardware accelerated, only to find out that it had been deprecated years ago.

Another reason I am interested, as a university student, is that I am a part of a program at Penn that funds and supports undergrad research interests. I have wanted to test whether exposure time to high frame rate content and the type (native or interpolated) leads to a preference for high frame rate content/media. There are many claims, even by movie directors, but no studies as far as I can tell. I don't mind this being public as I'm not doing this for a degree, but just as part of the program. I was going to use SVP and get funds for a test rig with an NVIDIA GPU (SVP takes advantage of NVIDIA's Optical Flow API), but if there is an implementation by Intel that comes out before next summer, I will be able to test the perceived preferences between both SVP and Intel.

But really, this is a personal interest of mine, and the research component would put my hypothesis that _well-done_ frame interpolation is preferable to native, lower frame rate content. If you want to read more, there is info on SVP's site (https://www.svp-team.com/) as well as their manual (https://www.svp-team.com/wiki/Manual:SVP). As for details about how this is normally achieved, the developers of mvtools have documentation available (http://avisynth.org.ru/mvtools/mvtools2.html).

The best way to understand it is to just try SVP and judge for yourself. I haven't gone into detail about how this can be applied for slow-motion video, but NVIDIA's research in this field is quite interesting: https://news.developer.nvidia.com/transforming-standard-video-into-slow-motion-with-ai/.
